### ISSUE TITLE
Adding lookup table support

### ISSUE THREAD

**Issue by Iluvmagick** [NilFoundation/zkLLVM#340]
After the circuit is generated, lookup tables need to be packed into the constant columns.
In the blueprint tests this is currently done like [this](
https://github.com/NilFoundation/zkllvm-blueprint/blob/1e40ce2035346feeac7f56c3463006507dfa7c5c/test/test_plonk_component.hpp#L291-L299).
The code needs to be called after the circuit generation is complete -- the circuit object should be keeping track of all the required lookup tables.
This requires https://github.com/NilFoundation/crypto3-zk/pull/198 from zk to work. The current packer version is prone to crashing if it is unable to fit the table into the provided constant columns/row amount -- I am working on fixing that ASAP.

EVM verifier currently supports lookup argument.

**Comment by Iluvmagick** [comment]
Also [this](https://github.com/NilFoundation/zkllvm-blueprint/blob/master/test/test_plonk_component.hpp#L202-L207) part is required in the sha component handling, as there is no proper manifest support for this yet.

**Comment by Iluvmagick** [comment]
For correct multi-prover support, the lookup tables need to be properly distributed between provers.
I see two different approaches for that.
Both require the lookup table packer to be called  **only once** after the entire circuit generation is complete! This is a **must**, otherwise the packer (correctly) crashes. In both cases the `fill_constant`/`fill_selector` methods in the proxy need to be removed, as they become redundant. The table reservation code still needs to be called at the spot where the sha component is placed. 

Approach one involves tracking which table exactly is allocated where during the table packing process. This requires modifying the table packing code to return the location of the packed tables. We then call the lookup table packer on the original (un-proxied) circuit and assignment objects. 
After the call we should use the information returned from the packer and the fact that the proxies track which lookup gates they use to get a set of all lookup tables used by each proxy and assign to each proxy the rows for their respective tables.
This also requires changing the proxies row-tracking code: currently it uses an `std::set` for that, which is extremely inefficient in practice for hundreds of thousands of rows. 
Using `std::unordered_set` would be better, but ideally we should take advantage of the fact that we do not deal with tracking collections of arbitrary rows here: in practice each proxy would likely have only a small amount of large contiguous intervals of rows. Use of a proper data structure would significantly speed both this step and also reduce the overhead in circuit generation in general.

The second approach requires instead packing the lookup tables into a **separate** assignment object. **All** other arguments to the packer should remain the same. We then use the value returned from the packer to ascertain the rows required for the table, and save all that is assigned into a shared column. 
This requires including additional shared columns in the assignment, and additional code to correctly append the shared columns to another place.
A significant upside of this is that the packer does not have to be touched at all (and currently we have no non-trivial tests for it besides running sha256 for ten minutes by using the parallel prover on the server), and this is likely to be faster in implementation overall.
A significant downside is that there is no way to implement table separation for different provers by using this approach, as there is no data available to do that (and the tables are **big**). 
This could also be remedied if we do follow a part of the first approach and modify the packer to be able to return the coordinates of different tables, but in that case we should probably go with the first approach altogether as most of the work for it would be already done.

**Comment by akokoshn** [comment]
Working on approach 2:
![image](https://github.com/NilFoundation/zkLLVM/assets/8722837/dc93cdca-3eb9-49aa-b66c-a4f5d01c1346)
Dev branch: https://github.com/NilFoundation/zkLLVM/tree/340-lookup-support-2
